url64700e9ab7_dwicke_course_scheduler_tgz-pJ8-org_json_TestJ8/src/coursescheduleapp/model/MGA.java

```diff
@@ -0,0 +1,477 @@
+     * fit out of the populaton.
+     *
+     * @param maxSeconds
+     * The maximum number of seconds the algorithm will run.
+     * @param popSize
+     * The number of individuals to create in each generation.
+     * @param findMax
+     * When true runGA will return the individual with max fitness
+     * If false runGA will return the individual with minimum fitness
+     * @return
+     * The individual based on findMax parameter.
+     */
+    public E runGAByTime(long maxSeconds, int popSize, boolean findMax, int numProc) {
+        this.popSize = popSize;
+        pool = java.util.concurrent.Executors.newFixedThreadPool(numProc);
+        createInitialPop();
+        System.gc();
+
+        long start = System.currentTimeMillis();
+
+        System.gc();
+        long stop = System.currentTimeMillis();
+
+
+        while ((stop - start) / 1000 <= maxSeconds) {
+            evalPopThreaded();
+            currGen = createGen();
+            whichGen++;
+            System.gc();
+            stop = System.currentTimeMillis();
+        }
+
+        return getResult(findMax);
+
+    }
+
+    /**
+     *
+     * This method will run the MGA and return E the individual that is the best
+     * fit out of the populaton.
+     *
+     * @param goalFitness
+     * The fitness goal number.
+     * @param maxGen
+     * Maximum number of generations the algorithm will complete
+     * @param popSize
+     * The number of individuals to create in each generation.
+     * @param findMax
+     * When true runGA will return the individual with max fitness
+     * If false runGA will return the individual with minimum fitness
+     * @return
+     * The individual based on findMax parameter.
+     */
+    public E runGAByGoal(double goalFitness, int maxGen, int popSize, boolean findMax) {
+        this.popSize = popSize;
+
+        createInitialPop();
+
+        goalFit = goalFitness;
+
+
+        while (goalFound == false && whichGen <= maxGen) {
+            evalPop();
+            if (goalFound == true) {
+                break;
+            }
+            currGen = createGen();
+            whichGen++;
+
+        }
+        return getResult(findMax);
+    }
+
+    private void createInitialPop() {
+        for (int i = 0; i < popSize; i++) {
+            currGen.add(individual.makeIndividual());
+        }
+    }
+
+    private E getResult(boolean findMax) {
+        int index = 0;
+
+         Entry<Integer, Double> result = null;
+
+        if (findMax == true) {
+           
+            Iterator<Entry<Integer, Double>> s = fitScores.entrySet().iterator();
+
+            
+           for (; s.hasNext(); )
+           {
+               Entry<Integer, Double> curr = s.next();
+               if (result == null)
+               {
+                   result = curr;
+               }
+               else if (curr.getValue() > result.getValue())
+               {
+                   result = curr;
+               }
+
+           }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 240:252, 258:270.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.